Direct solver with the full domain's grids

roelof-groenewald opened the pull request:
The implementation is complete, the PR is marked as draft since a unit test is still missing.
Example comparison showing implementation accuracy:
![solver_compare](https://user-images.githubusercontent.com/40245517/124995634-7d304480-dffc-11eb-9698-aadff2b6afcd.png)

roelof-groenewald committed:
added function to gather electrostatic potential for the entire domain on the root processor, only partially works

nruof committed:
First set of necessary changes when cherry-picking first commit from direct_solver branch

nruof committed:
Additional changes to make capacitive_discharge.py to run on this branch

roelof-groenewald committed:
changed getGatheredPhi_fp to return a reference to the full_phi_fp array, which seems to have fixed the issue with corrupted values in the array

peterscherpelz committed:
Initialize full_phi_fp similar to phi_fp

This seems to fix the bad values that were coming into full_phi_fp in
previous runs on > 1 processor on Penning. I wonder if somehow the
memory was not appropriately reserved in the previous initialization?

peterscherpelz committed:
Initial, totally-not-complete, diode setup script

roelof-groenewald committed:
added functions to gather rho on the full domain and return a pointer to that array

roelof-groenewald committed:
Merge branch 'nruof/get_grid_quanties' into direct_solver_full_grids

Commit:
Renamed Ar to inert_gas, raised not implemented errors, removed Cs-ionization functionality

roelof-groenewald committed:
added file poisson_pseudo_1d.py - the direct solver using superLU

roelof-groenewald committed:
bug fix in installfieldsolver functionality

Commit:
set up grids in init_base()

Commit:
Made not implemented errors more helpful

Commit:
Used PICMI electrostatic solver in init_solver()

Commit:
Boundry condition parameter fix and MAX_GRID_SIZE_FACTOR field

Commit:
Turned two fields into class input variables and added init_warpx() function

- Turned diag_steps and total_timesteps into class input variables called
  DIAG_STEPS and TOTAL_TIMESTEPS.

KZhu-ME committed:
Added functionality to set phi grid values from python script.

KZhu-ME committed:
Refactored comments and docstrings.

Commit:
commented out unused gentools.SetupCalcs() function

KZhu-ME committed:
Updated get phi grid to use get_pointer_full_phi_fp

Commit:
added sim_control variable

KZhu-ME committed:
Removed unnecessary return statements

KZhu-ME committed:
Merge pull request #11 from ModernElectron/kzhu/set_phi_grid

Added set phi grid functionality

Commit:
removed old imports

Commit:
Attempt at running capacitive_discharge using diode_setup

Commit:
Calcualte RES_LENGTH if NX, NY, NZ are passed in

roelof-groenewald committed:
cleaned up the direct solver some

Commit:
Species and field diagnostics now defined in diode_setup

- Particle species can be passed in as an input argument
- The list of field diagnostic data to be tracked can be defined as
  an input parameter.

Commit:
uncommented and started porting directory functions

Commit:
Passed dimension and rz to diode template, removed unused stuff

roelof-groenewald committed:
bug fix in direct solver to properly normalize rho

EtherealMC-Bit commented on Source/WarpX.cpp:
```diff
@@ -1487,13 +1491,40 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
+        Vector<int> pmap = {0};
+        dm_full.define(pmap);
+        // make a FabArray to hold the phi data
+        full_phi_fp[lev] = std::make_unique<MultiFab>(amrex::convert(ba_full,phi_nodal_flag),dm_full,ncomps,ngPhi,tag("full_phi_fp"));
```
Just curious why does `full_rho_fp` use a MultiFab with `2*ncomps` as a parameter while `full_phi_fp` uses one with just `ncomps`?

roelof-groenewald replied:
I have no idea. We just followed what was done in `rho_fp`, which also uses `2*ncomps`, but I don't know why.

EtherealMC-Bit commented on mewarpx/mewarpx/poisson_pseudo_1d.py:
```diff
@@ -0,0 +1,143 @@
+        """Function to build the superLU object used to solve the linear
+        system."""
+        self.nzsolve = self.nz + 1
+        self.nxsolve = self.nx + 3
```
Why the +1 and +3?

roelof-groenewald replied:
It creates the needed cells to enforce the specific boundary conditions. In the `z` direction we have Dirichlet boundaries so we only need one extra cell to fix the solution, while in the `x` direction we have periodic boundary conditions for which we need a ghost cell on each side and a third cell that is used as a dummy cell in order to enforce periodicity.

EtherealMC-Bit replied:
Interesting, thanks

EtherealMC-Bit reviewed (approved):
Cool stuff

peterscherpelz committed:
Significantly refactor testing and diode setup

peterscherpelz commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -1143,6 +1150,83 @@ def get_mesh_charge_density_fp(level, include_ghosts=True):
+def get_gathered_charge_density_fp(level):
+    '''
+
+    This returns a single numpy arrays containing the mesh charge density
```
arrays -> array

peterscherpelz commented on Python/pywarpx/_libwarpx.py:
```diff
@@ -1143,6 +1150,83 @@ def get_mesh_charge_density_fp(level, include_ghosts=True):
+def get_gathered_phi_fp(level):
+    '''
+
+    This returns a single numpy arrays containing the mesh electrostatic
```
arrays -> array

peterscherpelz commented on Python/pywarpx/callbacks.py:
```diff
@@ -256,7 +256,6 @@ def callfuncsinlist(self,*args,**kw):
 _c_afterdeposition = _CALLBACK_FUNC_0(_afterdeposition)
 libwarpx.warpx_set_callback_py_afterdeposition(_c_afterdeposition)
 _c_fieldsolver = _CALLBACK_FUNC_0(_fieldsolver)
-libwarpx.warpx_set_callback_py_fieldsolver(_c_fieldsolver)
```
Why did this line disappear?

roelof-groenewald replied:
If it is there `if (warpx_py_fieldsolver)` in `WarpXEvolve.cpp` evaluates to True even if a field solver is not installed since `warpx_py_fieldsolver` is no longer None. This is a bugfix in the implementation.

peterscherpelz commented on Source/WarpX.H:
```diff
@@ -273,6 +275,30 @@ public:
     const amrex::MultiFab& getBfield_avg_fp (int lev, int direction) {return *Bfield_avg_fp[lev][direction];}
 
+    const amrex::MultiFab& getGatheredRho_fp (int lev) {
+
```
Extraneous newline here? At least it's not consistent with below functions you added.

peterscherpelz commented on mewarpx/examples/capacitive_discharge.py:
```diff
@@ -214,11 +222,64 @@
 
-# step the simulation to the point where output should start
-sim.step(max_steps)
+def direct_solve():
```
Need comments on what this function is doing, or if you don't use it here remove it and the other commented-out changes.

peterscherpelz commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -151,6 +154,10 @@ def get_dt(self):
 
+    def get_t(self):
+        """Return the simulation time."""
+        return (self.get_it() - 1.0) * self.get_dt()
```
I assume you've checked that WarpX step 1 has time = 0? I'm too lazy to right now...

roelof-groenewald replied:
Haha I did.

peterscherpelz commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -174,5 +181,53 @@ def get_npart_species_dict(self):
+
+        Returns:
+            A list with only 1 element - a numpy array with rho on the full
+            domain.
```
Should clarify what this does on the other procs.

peterscherpelz commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -174,5 +181,53 @@ def get_npart_species_dict(self):
+
+        Returns:
+            A list with only 1 element - a numpy array with phi on the full
+            domain.
```
Same here, what happens on the other procs?

peterscherpelz commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -174,5 +181,53 @@ def get_npart_species_dict(self):
+            phi_ptr = _libwarpx.get_pointer_full_phi_fp(self.lev)
+            try:
+                phi_ptr[0][:] = phi_data
+            except ValueError:
```
You should check text of the ValueError to make sure this is indeed the error. Then, you should probably print what you believe is the problem and re-raise the error (just write `raise` on its own line), rather than SystemExit.

(Reasoning: In principle, you might have a python script that runs a bunch of WarpX simulations. It would have its own error handler that would mark a run as failed if it raises an error, but we wouldn't want this error to quit the script running all the simulations. So raising is almost always the better option - if nothing catches the error it'll cause program exit anyway.)

roelof-groenewald replied:
I'll change this now, but for learning sake @KZhu-ME see the above.

KZhu-ME replied:
I'll keep that in mind for the future. Thanks for the heads up!

peterscherpelz commented on mewarpx/mewarpx/poisson_pseudo_1d.py:
```diff
@@ -0,0 +1,143 @@
+        left_voltage = self.left_voltage(mwxrun.get_t())
+        right_voltage = self.right_voltage(mwxrun.get_t())
+
+        rho = -self.rho_data[
```
While it reads nicer to have this as a separate line, I think we should combine it with the source assignment. That avoids creating a whole new array from this division. eg
`source[1:-1,:] = (-self.dx**2 / constants.ep0) * self.rho_data[self.nxguardrho:-self.nxguardrho, self.nzguardrho:-self.nzguardrho]`

peterscherpelz replied:
You'll need to rewrite the nx, nz line but that's a simple arithmetic after getting self.rho_data.shape I think.

roelof-groenewald replied:
I hear you on the efficiency of not creating a new array but I think I'll leave it like this for now since it is the same way as we've been doing it in metools (where it hasn't proved to be a bottleneck) and more importantly since I think it keeps the workflow much cleaner in terms of array shapes. If you feel strongly about it I can certainly be convinced to change it though.

peterscherpelz replied:
Ah, no, I didn't realize it was already split out in metools. We'll see if it's taking time when we actually benchmark it then.

peterscherpelz reviewed (commented):
Lots of small notes.

KZhu-ME committed:
Added MCC wrapper to facilitate initializing MCC in PICMI script

roelof-groenewald committed:
changed direct solver to use the z-directions as being along the gap, also changed the example capacitive_discharge.py to use this convention

KZhu-ME committed:
added newline to end of file

peterscherpelz commented on mewarpx/examples/capacitive_discharge.py:
```diff
@@ -228,10 +210,13 @@
+##########################
+
+anode_voltage = lambda t: VOLTAGE * np.sin(2.0 * np.pi * FREQ * t)
+# comment line below to use the multigrid solver
```
And you should also note you need to uncomment `warpx_potential_hi_z` above I assume?

roelof-groenewald committed:
changed requested by PS during PR review

peterscherpelz commented on mewarpx/examples/capacitive_discharge.py:
```diff
@@ -243,43 +228,33 @@
+    import glob
+    import yt
+
+    data_dirs = glob.glob('diags/diags*')
```
I would check that data_dirs is not empty before proceeding, as otherwise rho_data division on line 259 will fail since that var was never created, which is a weird error when the problem is accessing data dirs.

roelof-groenewald commented:
@peterscherpelz Thanks for the comments. We'll address the requested changes to `capacitive_discharge.py` once @KZhu-ME's PR (#13) for the MCC wrapper is merged so that we can also change the example file for that functionality.

Commit:
fixed several issues

peterscherpelz committed:
Fix issues in moving from diode along x to along z

Also force PERIOD to be specified at present, and make RES_LENGTH always
reflect z direction.

peterscherpelz committed:
Make tests run in `temp_files`, not `test_files`.

KZhu-ME committed:
Added functionality to pass N_INERT or P_INERT and read from environmental variable, changed constants to be taken from picmi.py

peterscherpelz committed:
Remove old test no longer needed.

KZhu-ME committed:
Updated docstring and fixed path_name definition

Commit:
Initialized the simulation inside mwxrun instead of diode_setup

KZhu-ME committed:
Added error if file isn't recognized, and fixed units of N_INERT calculation

peterscherpelz committed:
Remove passing simulation to init_run.

Also fixed test_capacitive_discharge.py for full compatibility with
pytest, and check that a string from run completion is present.

roelof-groenewald committed:
Update mewarpx/mewarpx/mcc_wrapper.py

Co-authored-by: peterscherpelz <31747262+peterscherpelz@users.noreply.github.com>

roelof-groenewald committed:
Merge pull request #13 from ModernElectron/kzhu/feature_mcc-wrapper

Feature MCC Wrapper

Commit:
Suggested changes

- Initialized the picmi simulation with verbosity=0
- Defined variables for NX and NZ instead of reusing the numbers

roelof-groenewald committed:
Merge pull request #10 from ModernElectron/mkps/diode_test_template

Mkps/diode test template

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster' into direct_solver_full_grids

KZhu-ME committed:
Incorporated MCC wrapper in diode setup, added exclude_collisions argument

KZhu-ME committed:
Fixed indexing into requires_energy dictionary

KZhu-ME committed:
Added check if at least 1 scattering process is included, and fixed typos

roelof-groenewald committed:
updated example script for capacitive discharge to use the MCC wrapper

roelof-groenewald committed:
refactored direct solver to be a child of picmi.ElectrostaticSolver and added a partial test for it - still requires reference data to compare to

roelof-groenewald committed:
Merge remote-tracking branch 'public/memaster2' into direct_solver_full_grids

roelof-groenewald committed:
direct solver test complete with reference data from MGML solver to compare to

roelof-groenewald committed:
added warpx-data checkout command to circleCI config file

roelof-groenewald committed:
added pandas as a requirement for mewarpx

roelof-groenewald committed:
one directory up from warpx is three directories up from mewarpx_dir

roelof-groenewald committed:
turn ME_SOLVER on for circleCI builds

roelof-groenewald committed:
properly set ME_SOLVER to on

roelof-groenewald committed:
fixed the name of the direct solver test

peterscherpelz commented on mewarpx/mewarpx/setups_store/diode_setup.py:
```diff
@@ -381,6 +383,11 @@ def init_solver(self):
+                self.solver = poisson_pseudo_1d.PoissonSolverPseudo1D(
+                    grid=self.grid
+                )
+                return
```
I would make the code below an `else` clause rather than using `return` here. If someone happens to add more to this function below it, it seems counter-intuitive for that code to only apply to the multigrid case.

peterscherpelz commented on mewarpx/examples/capacitive_discharge.py:
```diff
@@ -36,37 +41,38 @@
+xmax = D_CA / nz * nx
+zmax = D_CA
 
 number_per_cell_each_dim = [32, 16]
```
Is it worth changing this to `[16, 32]` given the other x vs z changes?

peterscherpelz commented on mewarpx/mewarpx/testing_util.py:
```diff
@@ -67,3 +70,120 @@ def change_to_warpxdir(wd):
+
+    Returns:
+        minwmargin (float): If min < 0, this is min*(1-margin). If min > 0, this
+            is min*(1+margin).
```
Sphinx syntax this shouldn't be indented (eg `is` is even with `minwmargin`). Multi-line arguments do get indented because it's needed to identify when the next argument starts, Returns doesn't (at least last I used it). *shrug*.

roelof-groenewald replied:
This was copied directly from metools.util so we should keep in mind to fix it there then as well.

peterscherpelz commented on mewarpx/mewarpx/testing_util.py:
```diff
@@ -67,3 +70,120 @@ def change_to_warpxdir(wd):
+
+    Returns:
+        maxwmargin (float): If max < 0, this is max*(1-margin). If max > 0, this
+            is max*(1+margin).
```
same as above

peterscherpelz commented on mewarpx/mewarpx/testing_util.py:
```diff
@@ -67,3 +70,120 @@ def change_to_warpxdir(wd):
+
+    else:
+        print(("Record file {} does not exist. Reference data frame "
+               "not loaded.").format(record_file))
```
Would you mind converting this and below prints to f-strings? https://realpython.com/python-f-strings/ They're so magical I wish I had known about them earlier...

peterscherpelz commented on mewarpx/mewarpx/util.py:
```diff
@@ -4,9 +4,19 @@
 
 from pywarpx import geometry
 
+# CONSTANTS - SI
```
Let's put these in a new file `mwxconstants.py` unless you object... I came to regret having them all in `util` in `minerva`.

peterscherpelz replied:
I guess `util` will have to import `mwxconstants` at that point... that's probably OK though, `mwxconstants` should never need anything other than `pywarpx.picmi` and third-party packages so we don't get any import loops.

peterscherpelz commented on mewarpx/mewarpx/util.py:
```diff
@@ -46,3 +56,16 @@ def mkdir_p(path):
+
+
+def ideal_gas_density(p, T):
+    """Calculate neutral gas density from the ideal gas law.
```
I would add the Torr and cm^-3 units to this summary docstring so people are most likely to see it.

peterscherpelz commented on mewarpx/tests/test_direct_solver.py:
```diff
@@ -0,0 +1,86 @@
+"""Test full 1D diode run with diagnostics."""
```
Change docstring

peterscherpelz reviewed (commented):
Just some minor formatting and refactoring requests! Once resolved I'll be happy to approve.

roelof-groenewald committed:
changes requested by PS during PR review

PhilMiller commented on Source/WarpX.cpp:
```diff
@@ -1487,13 +1491,40 @@ WarpX::AllocLevelMFs (int lev, const BoxArray& ba, const DistributionMapping& dm
+        Vector<int> pmap = {0};
+        dm_full.define(pmap);
+        // make a FabArray to hold the phi data
+        full_rho_fp[lev] = std::make_unique<MultiFab>(amrex::convert(ba_full,rho_nodal_flag),dm_full,2*ncomps,ngRho,tag("full_rho_fp"));
```
Down the line, if we want to get some or all of this code upstream, unconditional allocation of full-domain-sized arrays will probably have to be changed.
